### DISCUSSION TITLE
"Clamp to border" effect in WebGL

### DISCUSSION THREAD

**Discussion by computersarecool** [gfxfundamentals/webgl-fundamentals#396] [Q&A]
To my recent dismay I recently learned that WebGL does not supported a "Clamp To Border".
So, if you shrink UV coordinates (to scale an image for example) it is on you if you do not want the texture repeated in some fashion (clamped, mirrored whatever).

On the web you could just add a one pixel alpha border like [this answer](https://stackoverflow.com/questions/59567256/how-can-i-get-a-clamp-to-border-effect-on-an-image-loaded-texture-in-three-js) suggests but it seems there should be a cleaner way (technically for non-square images I think that would very slightly change the aspect ratio).

Is the correct approach to check during your scaling step for UV coordinates past your scaling range?
I _think_ I am doing that correctly, but for textures that have zero whitespace, I still see a bit of clamping going on.

The situation that makes this harder is that I want to apply to a blur (which is why I am shrinking the texture) from a second texture, so I need valid UV coordinates in some of the scaled region (just not for the first texture).

Very curious here. Thanks!

**Comment by greggman** [comment]
If you want to simulate an OpenGL border I'm guessing this would work

    // ------
    // first compute a float "inside" where 0 = use borderColor
    // and 1 = use textureColor

    // 0 if < 0.0, 1 if >= 0.0
    vec2 insideBottomLeft = step(vec2(0.0), uv);

    // 1 if < 1.0, 0 if >= 1.0
    vec2 insideTopRight   = step(uv, vec2(1.0));

    // 1 if inside, 0 if outside
    vec2 insideBottomLeftTopRight = insideBottomLeft * insideTopRight;
    float inside = insideBottomLeftTopRight.x * insideBottomLeftTopRight.y;

    // -----
    // now clamp the UV coords so they don't bleed outside the edges of the texture

    // compute half a texel in UV coords
    vec2 halfTexel = 0.5 / textureResolution;
    vec2 minUV = halfTexel;
    vec2 maxUV = 1.0 - halfTexel;
    uv = clamp(uv, minUV, maxUV);

    vec4 textureColor = texture2D(tex, uv);

    vec4 color = mix(borderColor, textureColor, inside);

https://jsgist.org/?src=a11a5fbb89548ebcb0a97150421ede00


PS: I'm assuming OpenGL does not filter the texture with the border but I'm too lazy to go look it up.

PPS: the 2nd half of that shader, the part that clamps the UV coords, is not needed if you set the texture filtering to `gl.CLAMP_TO_EDGE`
